Add export form data to CSV sample

datalogics-bhaugen opened the pull request:
#### Changes in this pull request

- adds a sample that demonstrates exporting form data to a CSV file

#### No JIRA card

#### Checklist for approving this pull request

- [ ] There are **unit tests** for new/changed code, or a good explanation why this was not possible.
- [ ] All **CI builders** have indicated success (Give them a few minutes to notice the pull request.)
- [ ] The **Pull request title** has the JIRA issue numbers separated by spaces (if any), a space, and then a short, but descriptive summary.
- [ ] **Commit messages** are well formed: [A note about Git commit messages](http://www.tpope.net/node/106)
- [ ] New public packages, classes, and methods are **documented**. (Strongly consider documenting private classes and methods.)

#### Blockers

Add a checkbox for yourself with your **&#64;name** to this list if you are holding this PR open for review. PR Shepherd, please hold off merging this PR until these are all checked:

- [x] _&lt;add your name here with an **@**>_

datalogics-bhaugen committed:
Add sample that demonstrates exporting form data to CSV

Writes out form data (field names and formatted values)
to a CSV file

datalogics-bhaugen committed:
ExportFormDataToCsv : update throws descriptions

run fix_throws_javadoc.sh to update the throws
Descriptions in the javadoc

datalogics-bhaugen committed:
ExportFormDataToCsv : update param descriptions

datalogics-bhaugen committed:
Add filled_acroform.pdf

A copy of acroform_fdf.pdf that has been filled in
With dummy data.

datalogics-bhaugen committed:
filled_acroform : add more formatting rules

Added formatting rules to

- zip code (formats as zip+4)
- phone number
- cost per box
- delivery

datalogics-bhaugen committed:
filled_acroform : remove default value for oranges

datalogics-bhaugen committed:
ExportFormDataToCsv : try to print a value for every field

Get the list of values and if it is not null, then
print the first value in the list. This is not ideal
for every possible case out there, this is just a
demonstration though.

datalogics-bhaugen committed:
ExportFormDataToCsv : provide default input/output

So that the sample runs without the user having to
supply any input or output files.

datalogics-devauto commented:
Build failed.

datalogics-devauto commented:
Build succeeded.

datalogics-bhaugen committed:
Update .gitignore to ignore .csv files

datalogics-kam commented:
Hm. Interesting.

CSV is trickier than you think. All kinds of trouble escaping values that have funny values, or working to meet various standards. Would you consider [Apache Commons CSV](https://commons.apache.org/proper/commons-csv/)?

datalogics-devauto commented:
Build failed.

datalogics-devauto commented:
Build succeeded.

datalogics-kam committed:
IDEA module: update dependencies

datalogics-kam committed:
ExportFormDataToCsv: Use commons-csv CSVPrinter to write out data

Ensures that the format and escaping is correct.

datalogics-kam commented:
See datalogics-bhaugen#1

datalogics-bhaugen committed:
Merge pull request #1 from datalogics-kam/export-form-data-to-csv

Export form data to csv

datalogics-devauto commented:
Build succeeded.

datalogics-devauto commented:
Build failed.

datalogics-bhaugen committed:
Add basic test for ExportFormDataToCsv sample

datalogics-devauto commented:
Build succeeded.

datalogics-devauto commented:
Build failed.

datalogics-bhaugen committed:
ExportFormDataToCsvTest : check CSV data

- check the number of records that were read in
- check the number of entries in a record against
the expected number
- check the value of entries in a record against
the expected values

datalogics-bhaugen commented:
Added a test for this sample. The only thing remaining that I would like to do before this gets merged is add it to the lite package.

datalogics-devauto commented:
Build failed.

datalogics-devauto commented:
Build succeeded.

datalogics-bhaugen committed:
ExportFormDataToCsv : add private constructor

Silences a warning from PMD

datalogics-bhaugen committed:
ExportFormDataToCsv : flip if statement conditions

To quiet PMD

datalogics-bhaugen committed:
ExportFormDataToCsv : add conditionals to log statements

datalogics-bhaugen committed:
Add ExportFormDataToCsv sample/test to lite package

datalogics-devauto commented:
Build failed.

datalogics-devauto commented:
Build failed.

datalogics-bhaugen committed:
Include commons-csv in lite pom

datalogics-devauto commented:
Build failed.

datalogics-devauto commented:
Build failed.

datalogics-bhaugen committed:
ExportFormDataToCsvTest : rename constant

To fit in line length requirements as well as to
be clearer about what the data is.

datalogics-bhaugen committed:
ExportFormDataToCsvTest : format change by eclipse

datalogics-devauto commented:
Build failed.

datalogics-devauto commented:
Build failed.

datalogics-bhaugen commented:
Trigger build

datalogics-devauto commented:
Build failed.

datalogics-devauto commented:
Build failed.

datalogics-bhaugen committed:
ExportFormDataToCsv : make class final

datalogics-devauto commented:
Build succeeded.

datalogics-devauto commented:
Build succeeded.

datalogics-bhaugen committed:
ExportFormDataToCsvTest : specify encoding of CSV

The sample specifies that the CSV file will be
written out as UTF-8 so the test should also expect
to read the CSV back in as UTF-8

datalogics-devauto commented:
Build succeeded.

datalogics-devauto commented:
Build succeeded.

datalogics-bhaugen commented:
Latest commit resolved the last PMD warning that was left, this is ready for review now!

datalogics-kam commented on src/main/java/com/datalogics/pdf/samples/forms/ExportFormDataToCsv.java:
```diff
@@ -0,0 +1,179 @@
+            Files.delete(outputFile.toPath());
+        }
+
+        final PrintWriter writer = new PrintWriter(outputFile, "UTF-8");
```
``PrintWriter`` implements ``AutoCloseable``, so use try-with-resources to close it``
(So does ``CSVPrinter``)

datalogics-bhaugen replied:
Fixed in 9bb9847

datalogics-kam commented on src/test/java/com/datalogics/pdf/samples/forms/ExportFormDataToCsvTest.java:
```diff
@@ -0,0 +1,83 @@
+            final Map<String, String> recordAsMap = record.toMap();
+            assertEquals("Record size should match expected CSV data size", expectedCsvData.size(),
+                         recordAsMap.size());
+            assertEquals("", recordAsMap, expectedCsvData);
```
Empty string for the explanation makes me wonder if this was something planned that didn't make it.

datalogics-bhaugen replied:
Fixed in 5387e7d02, thanks for pointing that one out!

datalogics-kam commented on src/test/java/com/datalogics/pdf/samples/forms/ExportFormDataToCsvTest.java:
```diff
@@ -0,0 +1,83 @@
+            assertEquals("Record size should match expected CSV data size", expectedCsvData.size(),
+                         recordAsMap.size());
+            assertEquals("", recordAsMap, expectedCsvData);
+            recordAsMap.clear();
```
This confuses me. Why clear the map if it's going away anyway?

datalogics-bhaugen replied:
Also fixed in 5387e7d

datalogics-kam reviewed (commented):
This is kind of a start; I'll take more look at it later.

Good looking sample. Nice and concise.

The only other thing is to make sure your commit messages are imperative, and I don't know how I feel about the spaces before colons in commit headings, but that's maybe just me.

datalogics-bhaugen committed:
ExportFormDataToCsvTest: add explanation for assert

- remove unnecessary clear of a map as well

datalogics-bhaugen committed:
ExportFormDataToCsv: use try with resources

datalogics-bhaugen committed:
ExportFormDataToCsv: wrap document interaction with try

To ensure that the document is closed correctly

datalogics-devauto commented:
Build succeeded.

datalogics-devauto commented:
Build succeeded.

datalogics-coreyy commented:
LGTM, tested locally
